File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
from collections import deque
import sys
sys.setrecursionlimit(10**9)
input = sys.stdin.readline

n, m = map(int, input().split())
grid = [[] for _ in range(n + 1)]
for _ in range(m):
    u, v = map(int, input().split())
    grid[u].append(v)
    grid[v].append(u)

visited = [False] * (n + 1)
visited[0] = True
def dfs(v):
    if not visited[v]:
        visited[v] = True
        for next in grid[v]:
            dfs(next)

answer = 0
for i in range(1, n + 1):
    if not visited[i]:
        dfs(i)
        answer += 1

print(answer)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lines = [];

rl.on("line", (input) => {
  lines.push(input.trim());
  if (lines.length === Number(lines[0].split(" ")[1]) + 1) {
    rl.close();
  }
});

rl.on("close", () => {
  const [n, m] = lines[0].split(" ").map(Number);
  const grid = Array.from({ length: n + 1 }, () => []);
  for (let i = 1; i < m + 1; i++) {
    const [x, y] = lines[i].split(" ").map(Number);
    grid[x].push(y);
    grid[y].push(x);
  }

  const visited = new Set();
  function dfs(node) {
    const stack = [node];
    visited.add(node);

    while (stack.length) {
      const node = stack.pop();
      for (const next of grid[node]) {
        if (!visited.has(next)) {
          visited.add(next);
          stack.push(next);
        }
      }
    }
  }

  let answer = 0;
  for (let i = 1; i < n + 1; i++) {
    if (!visited.has(i)) {
      dfs(i);
      answer++;
    }
  }
  console.log(answer);
});
